vnext/Microsoft.ReactNative.Cxx/JSValue.cpp

```diff
@@ -3,9 +3,324 @@
 
 #include "pch.h"
 #include "JSValue.h"
+#include <iomanip>
+#include <sstream>
+#include <string_view>
 
 namespace winrt::Microsoft::ReactNative {
 
+//===========================================================================
+// JSValue type conversion helpers.
+//===========================================================================
+
+namespace {
+
+struct NullConverter {
+  static constexpr char const *NullString = "null";
+};
+
+struct ObjectConverter {
+  static constexpr char const *JavaScriptString = "[object Object]";
+};
+
+struct StringConverter {
+  static std::string_view TrimString(std::string const &value) noexcept {
+    constexpr char const *WhiteSpace = " \n\r\t\f\v";
+
+    size_t start = value.find_first_not_of(WhiteSpace);
+    if (start == std::string::npos) {
+      return "";
+    }
+
+    size_t end = value.find_last_not_of(WhiteSpace);
+    return {value.data() + start, end - start + 1};
+  }
+
+  static std::ostream &WriteAsJsonString(std::ostream &stream, std::string const &value) noexcept {
+    auto writeChar = [](std::ostream & stream, char ch) noexcept->std::ostream & {
+      switch (ch) {
+        case '"':
+          return stream << "\\\"";
+        case '\\':
+          return stream << "\\\\";
+        case '\b':
+          return stream << "\\b";
+        case '\f':
+          return stream << "\\f";
+        case '\n':
+          return stream << "\\n";
+        case '\r':
+          return stream << "\\r";
+        case '\t':
+          return stream << "\\t";
+        default:
+          if ('\x00' <= ch && ch <= '\x1f') {
```

Comment:
>\x1f [](start = 37, length = 4)

(nit) add comment about what is this character?

Reply:
These are just special non-printable characters in the beginning if ASCII table. I am not sure if their name would have any meaning.

---
In reply to: [389199211](https://github.com/microsoft/react-native-windows/pull/4260#discussion_r389199211) [](ancestors = 389199211)